dev/com.ibm.ws.kernel.service/src/com/ibm/ws/kernel/server/internal/ServerEndpointControlMBeanImpl.java

```diff
@@ -0,0 +1,173 @@
+    java.lang.Object betaDependency;
+
+    /*
+     * inject the service that will allow pause and resume control of endpoints
+     *
+     */
+    @Reference(service = PauseableComponentController.class,
+               cardinality = ReferenceCardinality.MANDATORY,
+               policy = ReferencePolicy.STATIC,
+               policyOption = ReferencePolicyOption.GREEDY)
+    private PauseableComponentController pauseableComponentController;
+
+    public ServerEndpointControlMBeanImpl() throws NotCompliantMBeanException {
+        super(ServerEndpointControlMBean.class);
+
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see com.ibm.websphere.kernel.server.ServerEndpointControlMBean#pause(java.lang.String)
+     */
+    @Override
+    public void pause(String targets) throws MBeanException {
+        try {
+            if (targets == null | "".equals(targets))
+                pauseableComponentController.pause();
+            else
+                pauseableComponentController.pause(targets);
+        } catch (PauseableComponentControllerRequestFailedException ex) {
+            throw new MBeanException(ex);
+        }
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see com.ibm.websphere.kernel.server.ServerEndpointControlMBean#resume(java.lang.String)
+     */
+    @Override
+    public void resume(String targets) throws MBeanException {
+        try {
+            if (targets == null | "".equals(targets))
+                pauseableComponentController.resume();
+            else
+                pauseableComponentController.resume(targets);
+        } catch (PauseableComponentControllerRequestFailedException ex) {
+            throw new MBeanException(ex);
+        }
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see com.ibm.websphere.kernel.server.ServerEndpointControlMBean#isPaused(java.lang.String)
+     */
+    @Override
+    public boolean isPaused(String targets) throws MBeanException {
+        try {
+            if (targets == null | "".equals(targets))
+                return pauseableComponentController.isPaused();
+            else
+                return pauseableComponentController.isPaused(targets);
+        } catch (PauseableComponentControllerRequestFailedException ex) {
+            throw new MBeanException(ex);
+        }
+    }
+
+    /*
+     * (non-Javadoc)
+     *
+     * @see com.ibm.websphere.kernel.server.ServerEndpointControlMBean#listEndpoints()
+     */
+    @Override
+    public List<String> listEndpoints() throws MBeanException {
+        Collection<PauseableComponent> pauseableComponents = pauseableComponentController.getPauseableComponents();
+        System.out.println("pauseableComponents length" + pauseableComponents.size() + " " + pauseableComponents.toString());
+        ArrayList<String> endpoints = new ArrayList<String>();
+        // list of endpoints may change while iterating
+        synchronized (this) {
+            while (true) {
```

Comment:
Could we get hung here?  The list we are iterating through will not change since we have set it outside the loop, so it seems to be we could get in a situation where we keep encountering the same error since we are looping through the same list and get stuck

Reply:
See above reply.